### ISSUE TITLE
Discussionの内容をキャッチアップする

### ISSUE THREAD

**Issue by meltyyyyy** [meltyyyyy/ICR-2023#3]
[参考]
https://www.kaggle.com/code/raddar/icr-competition-analysis-and-findings/notebook]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/420750

**Comment by meltyyyyy** [comment]
### BNのカラムは年齢

年齢が高いと病気持ちの可能性が高い 
年齢によってクラスタリングした特徴量は有効な気がする

### 訓練データではBQがNoneならクラスは0で確定

汎化を狙うならBQがNoneの特徴量はモデルに入れないべきかもしれない

### EJのカラムの値によっていくつかのデータは定数値となる

<img width="403" alt="スクリーンショット 2023-07-03 8 36 33" src="https://github.com/meltyyyyy/ICR-2023/assets/81362789/48b033db-e303-4868-b60d-4fc6e4a57c85">

### 時間経過(Epsilonカラムの変化)によってクラスの分布が変化している

<img width="720" alt="スクリーンショット 2023-07-03 8 37 50" src="https://github.com/meltyyyyy/ICR-2023/assets/81362789/1d69a383-362a-4d88-ac82-e361e3f06732">

時系列で予測クラスの分布が変化しているため, 典型的にCrossValidationはしない方が良い

### クラス1の外れ値がある

除外しないとクラス1の予測精度に大きく影響が出る

[出典]
https://www.kaggle.com/code/raddar/icr-competition-analysis-and-findings/notebook

**Comment by meltyyyyy** [comment]
### コロナの影響

前提として

`Epsilon The date the data for this subject was collected. Note that all of the data in the test set was collected after the training set was collected.`

とあるから, 訓練データ収集後にテストデータの収集を開始している. 2020年の訓練データは参考にならない可能性が高い.
コロナになっていない健康な人間のみがテストに参加していると考えられるから

[参考]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/420750

**Comment by meltyyyyy** [comment]
### 過学習について

- 「過学習」について：モデルの最適化プロセスの一部として、過学習に到達するのは自然なこと。しかし、汎用性を無視してリーダーボード（LB）に直接フィットさせることはモデリングとは言えない。
- ローカルCVスコアとLBスコアの違い：このコンペティションでは、ローカルCVスコアよりも高いLBスコアを見ることが一般的で、これは過学習ではなく、LBに直接フィットすることと一致している。
- 公開LBスコアの問題点：公開LBスコアを直接最適化するモデルは、頑健性や汎用性がなく、プライベートLBでのパフォーマンスは良くない可能性が高い。
- モデリングのアドバイス：ローカルCV評価を行わない高スコアの公開ノートブックは避けるべき。これらのノートブックの目的は、良好なPublic LBスコアを生み出すことだけであり、モデルの汎用性は考慮されていない可能性がある。

[出典]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/419930

**Comment by meltyyyyy** [comment]
### 訓練とCVとLBスコアの向上のバランスをとる方法

このコンペティションのメトリックはBalanced loglossです。そのため、訓練中にクラスをバランスさせることが重要です。クラスをバランスさせるための方法は少なくとも4つあります。

**方法1 - クラスの重みを追加**
クラスをバランスさせる最初の方法は、正のクラスに重みを追加することです。このコンペティションの訓練データでは、正のクラスは17%しかありません。そのため、モデルを訓練する際には、正のクラスに4.7倍の重みを付けて、正のクラスが負のクラスと同じ重要性を持つようにすることができます。(最適な重みを見つけるためには、異なる重みを試してみてください)。

ここにAmbrosMが提供したいくつかの例を挙げてみます：

```
LGBMClassifier(class_weight='balanced')
XGBClassifier(scale_pos_weight=4.71)
CatBoostClassifier(auto_class_weights='Balanced')
LogisticRegression(class_weight='balanced')
LinearDiscriminantAnalysis(priors=[0.5, 0.5])
```

**方法2 - 負のクラスをダウンサンプリング（または正のクラスをアップサンプリング）**
クラスをバランスさせるための2つ目の方法は、すべての正のサンプルと負のサンプルの一部だけを使用して訓練することです。これはダウンサンプリングと呼ばれています。訓練中にすべての負の値を使用していないので、毎回異なるダウンサンプリングを行う多くのモデルを訓練するべきです。（つまり、バギング）

ここにGanzyが提供したいくつかの例を挙げてみます：

from imblearn.under_sampling import RandomUnderSampler
```
NUM_POS = df.Class.sum()
sampler = RandomUnderSampler(sampling_strategy={0: NUM_POS, 1: NUM_POS})
X_train, y_train = sampler.fit_resample(df[FEATURES], df['Class'])
```
私のRAPIDS SVCノートブックでも「ダウンサンプリング」の例を提供しています。CV 0.44 => CV 0.35 および LB 0.35 => LB 0.27

**方法3 - ポストプロセス**
まだ公開ノートブックでは示されていない第三の方法は、ポストプロセスです。これは次のように動作します。もしもっと重みを正のクラスに加えずにモデルを訓練した場合、その後次のコードで正のクラスをポストプロセスでブーストすることができます：

```
boost = 4.7 # あなたのモデルに異なる因子を試してみてください
odds = boost * preds / (1-preds)
preds = odds / (1+odds)
```
ここで、変数predsはクラス1の予測値のNumPy配列です。確率をオッズに変換し、そのオッズをスケールしてクラス1をより重要にします。最後に、オッズを再び確率に変換します。

**方法4 - バランスの取れた対数損失を使用**
コメントでBjörnが提案した4つ目の方法は、訓練中に基本的な対数損失関数の代わりにバランスの取れた対数損失関数を使用することです。その後、クラスの重みをすべて等しく1として元のデータをそのままにすることができます。（注：ここで言及しているのは、早期停止メトリック関数ではなく、損失関数（つまり、GBTの目的）です）


[出典]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/412507

**Comment by TanimotoRui** [comment]
あざます！外れ値、年齢カラム、BQNoneあたりはちゃんと処理しないとダメそうだね。（もちろん他のもだけど）
これlbで使われているデータ多分privateに使われない気がするから、割とちゃんとlbに騙されすぎずちゃんとprivate当てに行こう。

> このリーダーボードは、テストデータの約42％で計算されている。最終的な結果は残りの58％に基づいて算出されるため、最終的な順位は異なる可能性がある。

**Comment by meltyyyyy** [comment]
@TanimotoRui 

ちょっと待ってまだ途中だった笑

**Comment by meltyyyyy** [comment]
### カラムはほぼ全て整数データ

ほぼ全てのカラムは整数値を取るようだが, 生データはfloatになっているので変換処理が必要.
整数値にした方が奇数/偶数による特徴量の生成などができるため, 有用である. 

[出典]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/413198

**Comment by meltyyyyy** [comment]
### Postprocessing の危険性

まず、私のLGBMモデルのクロスバリデーションされた予測をプロットします。このモデルのcvスコアは0.165です。ダイアグラムは、すべての予測のうち0.35より小さいものには、クラス1のサンプルが2つ、残りはクラス0であることを示しています：

![image](https://github.com/meltyyyyy/ICR-2023/assets/81362789/8844f80e-0717-4304-bc02-4e4a8ab0cd70)


最も左の赤いドット（予測確率が最も低いクラス1のサンプル）はy_pred ≈ 0.029に位置しています。これは、最大0.028であるy_predをすべてゼロに設定することでスコアを改善できることを示唆しています。スコアは0.162に改善します：

![image](https://github.com/meltyyyyy/ICR-2023/assets/81362789/b1abc723-1761-4c54-832a-70934f2fbd41)

これまでのところ、良好です - あなた自身のモデルで同じことができます（おそらく異なる閾値を選択する必要があるかもしれません）、そして、予測の両極端で同じ原則を適用することができます。

残念ながら、テストセットでは、最も左の赤いドットはy_pred ≈ 0.029にはなりません。それはどこでも可能です。プライベートテストセットが私の検証データと同じで、最も左の赤いドットがわずかに左に移動して予測が0.022になると仮定しましょう。元のLGBMモデルのスコアはあまり変わりません（0.165から0.166になりますが）、後処理モデルのスコアは0.162から0.305に悪化します：

![image](https://github.com/meltyyyyy/ICR-2023/assets/81362789/173788ec-8e4c-4123-9cd3-74a5f89e40a2)
![image](https://github.com/meltyyyyy/ICR-2023/assets/81362789/a1f451cf-d9e6-43c4-8762-102563a840c4)

私たちは、一つのクラス1サンプルに対して完全に誤った確率の0.0を予測するために、惨憺たるスコアを得ます！私たちの後処理は、モデルを過度に自信過剰にし、モデルが間違っているときに、loglossは過信気味のモデルを容赦なく罰します。

[出典]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/414209

**Comment by meltyyyyy** [comment]
### Balanced Logloss について

自信満々に間違ったらロスが非常に大きくなる, 自信なくて間違っても低いロスで済む

<img width="318" alt="スクリーンショット 2023-07-03 23 42 29" src="https://github.com/meltyyyyy/ICR-2023/assets/81362789/507458ed-81d3-4958-8d10-0f4831c1173d">

![image](https://github.com/meltyyyyy/ICR-2023/assets/81362789/e6dc5e8d-f745-46be-b294-9c58123bc4bc)

[出典]
https://www.kaggle.com/code/dan3dewey/icr-2023-balanced-log-loss

**Comment by meltyyyyy** [comment]
### 欠損値について

**結論 : 欠損値は普通に埋めて良い**

データセットのサイズが小さい場合、データの欠損について心配するべきかどうかは正当な問いです。そして、もしそうであるなら、欠損値をどのように補完すべきでしょうか？

私の答えを述べる前に、テストデータセットで見られる欠損データの種類と量が訓練データと同じであると仮定します。その仮定のもとで、私たちは欠損データについて悩むべきではないと思います。ここで私が明確にしたいのは、欠損データをどのように補完するかは問題ではないということです。しかし、テストデータにはNaN値が存在する可能性が高いことについては、非常に懸念すべきだと思います。そして、訓練データと同じ方法でテストデータの何かを補完すべきです。私たちのノートブックがツリーベースの方法を使用している場合、補完なしでも問題ないかもしれません（これらの方法は内部でNaNを処理します）。しかし、一般化された線形方法やニューラルネットワーク、またはデータのスケーリングや正規化を行っている場合、NaNは何かに置き換えるべきです。これは訓練データとテストデータの両方に当てはまり、代理の"テスト"データセットが何も欠けていない場合でもです。欠損データを補完する方法は多くあり、最良の方法を見つけるのは皆さんに任せます。疑問がある場合、これは一時しのぎとして機能します（そして私はデータをプロットする前に同じことを行いました）：

```
train_df.fillna(train_df.mean(), inplace=True)
test_df.fillna(test_df.mean(), inplace=True)
```

以下は訓練データセットのt-SNEプロットです。私がこれを使用しているのは、データフレームを通して見るよりも、2Dで多くのデータポイントを見る方が簡単だと感じるからです。左上のパネルは、0と1のラベルを持つ点がどれほど分離できるかを示しています。その隣には、我々の二つのクラスに関して欠損値の分布があります。興味深い部分は下部です。ここでは、私のLightGBMモデルの一つに基づいて絶対誤差値で色分けされたポイントがあります。もし真のラベル値が1で予測が0.9であれば、エラーは0.1となり、そのデータポイントは青色になります。もし真のラベル値が0で予測が0.9であれば、絶対誤差は0.9となり、そのデータポイントは赤色になります。左下の角を見ると、ほとんどのデータポイントがうまく予測されていることがわかります。オレンジ色と赤色のドットは、0と1のポイントが混在している領域にあり、これは分類器がそれらを分離するのが難しいことを示しています。欠損値を持つデータポイントだけが表示されている右下の角を見ると、ほとんどのポイントが濃い青、青、または水色であることがはっきりとわかります。これは、それらが比較的小さな絶対誤差で予測されていることを意味します。

![image](https://github.com/meltyyyyy/ICR-2023/assets/81362789/84634208-5afb-4d6f-b2b6-b5d4497f9044)


[出典]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/410843

**Comment by meltyyyyy** [comment]
### 有用な情報まとめ

**Solution**

データセットを拡張し、問題をトップk探索問題に変換する新しい方法を提案しました。
最良の公開カーネルの一部は過学習している可能性があると懸念しています。
データのクロスバリデーションを行わず、歴史の法則や情報の未来の完全なドリフトを使用する高スコアの公開カーネルは危険であると述べています。

**データ処理**

異なる補間技術を用いてNaNを取り扱う：著者は欠損値の取り扱いに関するいくつかの方法を提供していますが、現在の競技会のためのコードは提供していません。

欠損値を持つデータポイントについて心配する必要はありません

著者の見解は、ツリーモデルを使用する際にはNaN値について心配する必要はないというものです。著者の立場は、欠損値をどのように埋めるかはあまり重要ではなく、埋めることができればそれでよいというものです。これは、欠損値を持つデータポイントが平均値で最も単純な補間を行った場合でもよく予測されることから推察されます。

🔍外れ値の特定ガイド📈🚀、著者は外れ値の特定方法を提供していますが、現在の競技会のためのコードは提供していません。

Zスコア、MAD、Tukeyのフェンス(IQR)、マハラノビス距離、ローカル外れ値因子(LOF)

この競技会のデータを対数変換することは可能ですか？

コメントセクション：はい、何でも試すことができます。しかし、GBTモデルではスプリットを使用しているため、対数変換はパフォーマンスに影響を及ぼしません。対数変換はNN、KNN、SVCなどの他の多くのモデルに影響を及ぼします。

私たちはデータセットをより魅力的にすることができますか？それは助けになりますか？（2023年6月27日更新）

著者は、視覚的に、データには高い区別度がないという問題を提起しています。区別度を高めるために一連の方法を使用する方がよいでしょうか？
著者は、一連の処理を通じて区別度を高める試みを行い、CVは大幅に改善しましたが、LBは減少しました。

データドリフトとEpsilonの使用について（2023年6月27日更新、2023年7月3日の個人的なまとめ）

y_labelだけでなく、multilabel（内部）も"データがシフトしている"。
著者は年月別のデータ数をプロットしており、2019年10月、2020年4月、2020年9月に3つの"穴"が見られ、これが"未知"のデータの出所であると推測しています。

**特徴生成 & 選択**

時間、つまりEpsilonを変数として使用する

著者は、多くの高スコアのカーネルがテストセットのEpsilonの値をeps_train_max + 1に設定していると述べています。しかし、同じロジックを使用して2020年のデータを分離し、オフラインでテストしてみたところ、結果は満足のいくものではなかったとのことです。

"ブルートフォース列選択"による過学習のリスク?

特徴検索に関連する問題について、著者のアプローチはOOF（Out Of Fold）+ Treemodel_featureimpループで、これは過学習になりやすいです。
コメントセクション
データポイントが約600点、特徴が56ある場合、特徴を増やすより減らすべきだと思います。少なくとも機械学習の目的にとっては、このデータセットは未定性が増しています。完全を期すため、遺伝的プログラミングを用いた特徴エンジニアリングを少なくとも10回分の実行、それぞれ1000世代を行いました。その結果、特徴は中性であったか、あるいは過学習につながることが多く、これは驚くべきことではありません。理論的には、シンプルなモデル、そして可能性としては特徴をスパース化する何かがこの問題の解決策であるはずです。それも試し、特徴を26-32に減らしたのですが、これまでのところそれも過学習につながっています。

特徴選択について（2023年6月27日更新）

著者は、データセットにノイズ特徴が存在すると信じており、これらのノイズ特徴を除去するためのソースコードを提供しています。コメントセクションでは、ユーザーが自身の特徴選択戦略を共有していますが、その中にはCVの向上とLBの減少につながるものもあるようで、これは一般的な状況のようです。

**CV**

適切なCV方法を使用することが非常に重要であるとし、多くのクロスバリデーションコードも提供しています。
ネストされたバリデーション戦略を採用することで過学習問題を追跡できると提案しています。

**その他**

この競技会で知っておくべきこととして、クロスバリデーションは無駄であり、リーダーボードは無意味であると主張しています。
メタデータの使用方法、ポストプロセスリングリスクの説明、以前の競技会からの有用な素材などについても触れています。

[出典]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/417983

**Comment by meltyyyyy** [comment]
### Balanced Loglossのコード

```
import numpy as np
def balanced_log_loss(y_true, y_pred):
  # calculate the number of observations for each class
  N_0 = np.sum(1 - y_true)
  N_1 = np.sum(y_true)
   # calculate the weights for each class
  w_0 = 1 / N_0
  w_1 = 1 / N_1
   # calculate the predicted probabilities for each class
  p_0 = np.clip(y_pred[:, 0], 1e-15, 1 - 1e-15)
  p_1 = np.clip(y_pred[:, 1], 1e-15, 1 - 1e-15)
   # calculate the log loss for each class

  log_loss_0 = -w_0 * np.sum((1-y_true) * np.log(p_0))
  log_loss_1 = -w_1 * np.sum(y_true * np.log(p_1))
   # calculate the balanced logarithmic loss
  balanced_log_loss = (log_loss_0 + log_loss_1) / (w_0 + w_1)
   return balanced_log_loss
```

[出典]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/409691

**Comment by meltyyyyy** [comment]
### EDAの投稿

非常に参考になる.

[出典]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/418474

**Comment by meltyyyyy** [comment]
### ベースラインの比較

<img width="755" alt="スクリーンショット 2023-07-04 8 30 44" src="https://github.com/meltyyyyy/ICR-2023/assets/81362789/cec6d2f4-4303-480a-9068-ed9c9083af7c">

[出典]
https://www.kaggle.com/competitions/icr-identify-age-related-conditions/discussion/411479

**Comment by meltyyyyy** [comment]
@TanimotoRui 

追加した分まで確認したらクローズしてもらって大丈夫です !

**Comment by TanimotoRui** [comment]
あざます！クローズします！

**Comment by TanimotoRui** [comment]
時系列で穴が観測されてそこがテストのデータセットになっている、の主張はすごいな、そんなこともわかるんかって思った
